spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/convert/SpannerConverter.java

```diff
@@ -82,4 +82,44 @@
 	 * are written.
 	 */
 	void write(Object source, WriteBuilder sink, Set<String> includeColumns);
+
+	/**
+	 * Determines if this converter can handle the given types for writing to Spanner
+	 * @param type the type of the property that data will be converted from Spanner in to
+	 * @param spannerSupportedType the desired Spanner supported type for reading.
+	 * @return true if the conversion is possible. false otherwise
+	 * @throws org.springframework.cloud.gcp.data.spanner.core.mapping.SpannerDataException if
+	 * the given desired Spanner supported type is not actually a supported Spanner type
+	 */
+	boolean canHandlePropertyTypeForSingularRead(Class type, Class spannerSupportedType);
```

Comment:
Do you mean to add the differences in these 4 functions as parameters to a single function?

Reply:
* have these as private functions on SpannerConverter
* expose a `canConvert` function that by the way you already introduced in another PR

The following piece of code in `SchemaUtils` knows too much about conversion and can be just turned into a `converter.canConvert(...)` call: 

````
if (isIterableInnerType
+					&& spannerConverter.canHandlePropertyTypeForArrayRead(originalType,
+							spannerType)
+					&& spannerConverter.canHandlePropertyTypeForArrayWrite(originalType,
+							spannerType)) {
+				ret = spannerType;
+				break;
+			}
+			else if (!isIterableInnerType
+					&& spannerConverter.canHandlePropertyTypeForSingularRead(originalType,
+							spannerType)
+					&& spannerConverter.canHandlePropertyTypeForSingularWrite(
+							originalType, spannerType)) {
+				ret = spannerType;
+				break;
+			}

````

Reply:
The "forArray" vs "forSinglular" distinction is important because Spanner can't handle the same types for array as for singular objects. For example, `double[]` can be a singular write but cannot be a an array write. I could combine the two forArrays and the two forSingulars, though. But I would need to name them "canConvertReadAndWriteForArray" and "canConvertReadAndWriteForSingular" to show that we can convert in both directions. How does that sound?

Reply:
We can still have the 4 methods as private methods. Why do you need them to be non-private?

Reply:
I am trying to simplify the API - the narrower the surface the better - makes it easier to implement for users for example if they wanted to extend it. I see that currently introducing these finer grained methods on the Converter interface is not justified.

Reply:
good point! changed